### DISCUSSION TITLE
클로저에 대해서 설명해주세요

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#11] [프론트 기술 질문 공간]
실행컨텍스트 가지고 설명해주세요

**Comment by javascriptbiggosoo** [comment]
클로저 덕분에 함수는 그 함수가 만들어진 곳에 있었던 변수를 계속 쓸 수 있습니다.

```jsx
const secureBooking = function () {
  let passengerCount = 0;
  return function () {
    passengerCount++;
    console.log(`${passengerCount} passengers`);
  };
};

const booker = secureBooking();
booker(); // 1 passengers
booker(); // 2 passengers
booker(); // 3 passengers
```

? 이게 왜 될까요? 기존 상식대로라면 모두 '1 passengers'이 프린트 되어야 하는게 아닐까요?

`secureBooking`은 **EC**에서 이미 빠져나가고 function 안에 만들어진 variable은 global scope에서 못써야하지 않나요?

클로저 때문입니다.

클로저는 함수가 만들어진 곳에 있었던 변수를 계속 쓸 수 있게 해준다는 것.

클로저를 배우지 않은 상태에서 생각해본다면 booker에 함수를 할당하는 secureBooking()을 호출한 후에 함수는 더 이상 EC에 존재하지 않기 때문에 역할을 하지 못해야 할텐데, 실행해보면 여전히 기존 함수 내의 변수에 접근이 가능합니다.

JS에서는 함수가 속해있던 함수가 더이상 EC에 존재하지 않더라도 그 변수환경에 접근할 수 있습니다. 이것은 JS의 특성으로 그렇지 않은 언어들이 많습니다.

클로저: 함수가 만들어진 당시의 그 위치, 시점의 변수환경이 그 함수에 계속 붙어있는, 그 조합을 클로저라고 부릅니다.

클로저 스코프가 글로벌 스코프보다 우선순위가 높아서 만약 전역에 passengerCount = 10 이 있더라도 클로저의 0을 받게됩니다.

**Comment by michoball** [comment]
간략히 말하자면 함수가 선언된 시점에 렉시컬 환경을 기억하고 그 환경내의 변수들을 참조 하는 것을 말합니다.

자바스크립트에서만 있는 개념이 아닌 함수형 프로그래밍 언어에서 볼 수 있는 특징입니다. 

함수에 대한 실행 컨텍스트가 생성될 때 함수 스코프내에 변수에 대한 값을 찾을 수 없으면 이 함수가 선언된 시점에 렉시컬 환경을 타고 상위 스코프를 둘러보며 해당 값에 대한 참조 값을 찾습니다. 

그래서 상위 스코프 내에 선언된 변수를 하위 스코프 함수에서 사용할 수 있는 것 입니다. 

react에서는 이러한 특징을 이용해서 useState나 useEffect 등의 hook을 만들었습니다.
